src/clientHandler.ts

```diff
@@ -108,19 +108,64 @@ export class ClientHandler {
   private async getServerOptions(): Promise<ServerOptions> {
     const cmd = await this.lsPath.resolvedPathToBinary();
     const serverArgs = config('terraform').get<string[]>('languageServer.args', []);
+    const procEnv = this.getProcessEnvironment();
+
     const executable: Executable = {
       command: cmd,
       args: serverArgs,
-      options: {},
+      options: {
+        env: procEnv,
+      },
     };
+
     const serverOptions: ServerOptions = {
       run: executable,
       debug: executable,
     };
+
     this.outputChannel.appendLine(`Launching language server: ${cmd} ${serverArgs.join(' ')}`);
     return serverOptions;
   }
 
+  private getProcessEnvironment() {
+    // deep object clone without the bother of custom traversal
+    const procEnv = JSON.parse(JSON.stringify(process.env));
+
+    // It's unlilely but possible that there is no PATH set. Object property
+    // names can be case sensitive depending on the OS, so we cannot rely on
+    // process.env.PATH working everywhere. It could be called PATH or Path or path
+    // and indexing would fail to find it.
+    if (procEnv.PATH === undefined) {
+      let envPath = '';
+      // Look through all of the environment names looking for PATH in a case
+      // insensitive way and remove the conflicting env var.
+      Object.keys(procEnv).forEach(function (keyname) {
+        if (keyname.match(/^PATH$/i)) {
+          envPath = procEnv[keyname] ?? '';
+          procEnv[keyname] = undefined;
+        }
+      });
+      procEnv.PATH = envPath;
+    }
```

Comment:
Is there any particular situation which you ran into that motivated you to add this?

I'm thinking that _if_ someone already has wrong casing, this must be causing various other problems, so I'm not sure we'd be helpful in hiding those problems in such a scenario?

Reply:
Yes, past experience with the puppet extension.

It's not 'someone' setting a 'wrong' casing. It's that the range of OS we support that each have different rules on casing and set key `PATH` casing differently. If we use `process.env.PATH` in code, but it is `process.env.Path` and access it on nix we won't get the value because the casing is different, where on windows we will get the correct value.

If we sanitize our inputs, we can at least be certain what is being passed down to the LS, which will aid debugging bug reports in the future.

Supporting documentation:

[child._process](https://github.com/nodejs/node/blob/master/doc/api/child_process.md):

> On Windows, environment variables are case-insensitive. Node.js lexicographically sorts the env keys and uses the first one that case-insensitively matches. Only first (in lexicographic order) entry will be passed to the subprocess. This might lead to issues on Windows when passing objects to the env option that have multiple variants of the same key, such as PATH and Path.

[process.env](https://nodejs.org/api/process.html#processenv)

>On Windows operating systems, environment variables are case-insensitive.
>
> import { env } from 'process';
> 
> env.TEST = 1;
> console.log(env.test);
> // => 1

Reply:
Hmm, OS differences, I see. 🤔 

It looks like Go stdlib assumes more explicitly that on Windows it's always all lowercase:
https://cs.opensource.google/go/go/+/refs/tags/go1.17.5:src/os/exec/lp_windows.go;l=87

Is there a reason why not follow that? It would probably simplify the code a bit and make it easier to read/understand.

Reply:
btw. I'm somewhat surprised the NodeJS stdlib doesn't handle this automatically.

Reply:
I can't speak for what is done in stdlib. I'm just trying to work around documented problems in accessing the value for PATH